working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1537,130 +1639,46 @@ To type check a pattern `p` being matched against a value of type `M`:
-the record because `obj` is known to be a record type of the right shape. But in
-the second function, `obj` may fail to match because the value may not be a
-record. *This implies that we can't determine whether a pattern in a variable
-declaration is incorrectly refutable until after type checking.*
+It is a compile-time error if:
 
-Refutability of a pattern `p` matching a value of type `V` is:
+*   The type of an expression in a guard clause is not `bool` or `dynamic`.
 
-*   **Logical-and**, **parenthesized**, **null-assert**, or **cast**:
-    Irrefutable if and only if all subpatterns are irrefutable.
+*   `p` is in an irrefutable context, it has a required type `T`, and `M` is not
+    assignable to `T`. *Destructuring and variable patterns can only be used in
+    declarations and assignments if we can statically tell that the
+    destructuring and variable binding won't fail.*
 
-*   **Logical-or**, **relational**, **null-check**, **literal**, or
-    **constant**: Always refutable. *Logical-or patterns are refutable because
-    there is no point in using one with an irrefutable left operand. We could
-    make null-check patterns irrefutable if `V` is assignable to its static
-    type, but whenever that is true the pattern does nothing useful since its
-    only behavior is a type test.*
+### Pattern uses
 
-*   **variable**, **list**, **map**, **record**, or **extractor**: Irrefutable
-    if and only if `V` is assignable to the static type of `p` and all
-    subpatterns are irrefutable. *If `p` is a variable pattern with no type
-    annotation, the type is inferred from `V`, so it is never refutable.*
+It is a compile-time error if the expression in a guard clause in a switch case
+or if-case statement is not assignable to `bool`.
 
-It is a compile-time error if a refutable pattern appears in an irrefutable
-context. *This means that the explicit predicate patterns like constants and
-literals can never appear in pattern variable declarations or pattern
-assignments. The patterns that do type tests directly or implicitly can appear
-in variable declarations or assignments only if the tested type is assignable
-from the value type. In other words, any pattern that needs to "downcast" to
-match is refutable.*
+The static type of a switch expression is the least upper bound of the static
+types of all of the case expressions.
 
 ### Variables and scope
 
 Patterns often exist to introduce new variable bindings. A "wildcard" identifier
-named `_` in a variable or cast pattern never introduces a binding.
-
+named `_` in a variable pattern never introduces a binding.
 The variables a patterns binds depend on what kind of pattern it is:
 
 *   **Logical-or**: Does not introduce variables but may contain subpatterns
-    that do. If it a compile-time error if the two subpatterns do not introduce
+    that do. It is a compile-time error if the two subpatterns do not introduce
     the same variables with the same names and types.
```

Comment:
Yeah, I agree this is definitely too vague. For now, I'm going to just put a "TODO" here, because I don't want to bother specifying something in detail that will then likely get replaced when I address https://github.com/dart-lang/language/issues/2473.